### PR TITLE
Allow NULLs in CASE and transform

### LINKED ISSUE #7237: CASE  unexpected behaviour
```
SELECT id,  
    CASE id 
         WHEN 1 THEN 'Z'  
    END x 
FROM  (SELECT 1 as id)

┌─id─┬─x────┐
│  1 │ ᴺᵁᴸᴸ │
└────┴──────┘

SELECT id,  
       CASE id 
            WHEN 1 THEN 'Z' 
            ELSE 'X' 
     END x 
FROM  (SELECT 1 as id)

┌─id─┬─x─┐
│  1 │ Z │
└────┴───┘
18.14.18, 19.16.1.1414 
```

### PR DESCRIPTION
**Pull request by aiven-sal**
`CASE` uses `caseWithExpression` behind the scenes, which in turn calls `transform`.
To be able to use `CASE` without an `ELSE` branch we need `transform` to be able to deal with NULL default values.

To achieve this I completely rewrote the transform function. Some old tests are broken by these changes, but it looks like those tests were wrong in the first place: what they are testing doesn't seem to be coherent with the documented behaviour of `transform`.

Closes #7237

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix a bug that prevented the use of `CASE` without an `ELSE` branch and extended `transform` to deal with more types. Also fix some bugs that made transform() return incorrect results when decimal types were mixed with other numeric types.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: tests/queries/0_stateless/02497_if_transform_strings_to_enum.sql [hunk 1]
```diff
@@ -33,13 +33,13 @@ SELECT transform(number, [2, 4, 6], ['google', 'censor.net', 'yahoo'], 'other')
...
 
-SELECT transform(number, [NULL], ['google', 'censor.net', 'yahoo'], 'other') FROM (SELECT NULL as number FROM system.numbers LIMIT 10);
+SELECT transform(number, [NULL], ['google', 'censor.net', 'yahoo'], 'other') FROM (SELECT NULL as number FROM system.numbers LIMIT 10); -- { serverError 36 }
 EXPLAIN SYNTAX SELECT transform(number, [NULL], ['google', 'censor.net', 'yahoo'], 'other') FROM (SELECT NULL as number FROM system.numbers LIMIT 10);
 EXPLAIN QUERY TREE run_passes = 1 SELECT transform(number, [NULL], ['google', 'censor.net', 'yahoo'], 'other') FROM (SELECT NULL as number FROM system.numbers LIMIT 10);
 
-SELECT transform(number, NULL, ['google', 'censor.net', 'yahoo'], 'other') FROM system.numbers LIMIT 10;
-EXPLAIN SYNTAX SELECT transform(number, NULL, ['google', 'censor.net', 'yahoo'], 'other') FROM system.numbers LIMIT 10;
-EXPLAIN QUERY TREE run_passes = 1 SELECT transform(number, NULL, ['google', 'censor.net', 'yahoo'], 'other') FROM system.numbers LIMIT 10;
+SELECT transform(number, NULL, ['google', 'censor.net', 'yahoo'], 'other') FROM system.numbers LIMIT 10; -- { serverError 43 }
```
Comment (antonio2368) [inline tests/queries/0_stateless/02497_if_transform_strings_to_enum.sql:40]: This seems wrong, you are throwing exception when it failed to apply some optimization to a single transform.
What if you have in the same query `transform` for which this optimization can be applied?
Better fix IMO would be to modify `ConvertStringsToEnumVisitor` and `IfTransformStringsToEnumPass`.

Comment (antonio2368) [reply tests/queries/0_stateless/02497_if_transform_strings_to_enum.sql:40]: ah yes that also makes sense 👍

**Comment by CurtizJ** [comment]
Unfortunately you cannot remove all specializations and leave only generic version. The performance will degrade significantly. You have to add a generic version in addition.

Manually tested query from performance tests:
```
clickhouse benchmark <<< "SELECT transform(number, [2, 4, 6], ['google', 'yandex', 'yahoo'], 'other') as x FROM numbers(10000000) GROUP BY x FORMAT Null" --cumulative
```

Your PR:
```
localhost:9000, queries 63, QPS: 6.843, RPS: 68478768.709, MiB/s: 522.452, result RPS: 0.000, result MiB/s: 0.000.

0.000%          0.143 sec.
10.000%         0.143 sec.
20.000%         0.144 sec.
30.000%         0.144 sec.
40.000%         0.145 sec.
50.000%         0.145 sec.
60.000%         0.146 sec.
70.000%         0.147 sec.
80.000%         0.148 sec.
90.000%         0.150 sec.
95.000%         0.152 sec.
99.000%         0.153 sec.
99.900%         0.155 sec.
99.990%         0.155 sec.
```

Current master:
```
localhost:9000, queries 91, QPS: 9.530, RPS: 95372740.442, MiB/s: 727.636, result RPS: 0.000, result MiB/s: 0.000.

0.000%          0.100 sec.
10.000%         0.101 sec.
20.000%         0.102 sec.
30.000%         0.102 sec.
40.000%         0.102 sec.
50.000%         0.103 sec.
60.000%         0.103 sec.
70.000%         0.103 sec.
80.000%         0.104 sec.
90.000%         0.105 sec.
95.000%         0.112 sec.
99.000%         0.155 sec.
99.900%         0.158 sec.
99.990%         0.158 sec.
```

The difference is about 40%.

**Comment by CurtizJ** [comment]
> Do you remember exactly what commits did you test?

No, it was the latest master. Anyway let's wait the result of perf tests in CI.

**Review by CurtizJ — APPROVED** [review]
LGTM

**Comment by CurtizJ** [comment]
Please, resolve the conflicts and I'll merge the PR then.

**Review by CurtizJ — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@CurtizJ Unfortunately, the Fuzzer has found a bug in this pull request:
```
milovidov-desktop :) SELECT caseWithExpr(arrayReduce(NULL, []), []);

SELECT caseWithExpr(arrayReduce(NULL, []), [])

Query id: 24445c4f-bb82-49dc-b025-8bc5c84fde6b


0 rows in set. Elapsed: 0.001 sec. 

Received exception:
Code: 49. DB::Exception: Value of type Array(Nothing) in memory is not of fixed size.: While processing caseWithExpr(arrayReduce(NULL, []), []). (LOGICAL_ERROR)
```

**Comment by alexey-milovidov** [comment]
This PR has bad code style: cryptic variable names.

`dt` -> data_type
`dtsize` -> data_type_size

**Comment by alexey-milovidov** [comment]
I will help...

**Comment by alexey-milovidov** [comment]
#49616

**Comment by alexey-milovidov** [comment]
#49625

**Pull request merged**